### PR TITLE
[OpenVINO-EP] Enabling save/Load blob feature

### PR DESCRIPTION
**Pull request by MaajidKhan**
**Description:** Enabling save/Load blob feature for OpenVINO-EP

Note: 
1.This feature is currently enabled only for VPU (MyriadX Device) both for Linux and windows. we would be enabling it for keembay as well in future.
2.This feature is not supported for other plugin's like CPU, GPU from OpenVINO.

**Motivation and Context:**
**OPTION 1:** (currently whatever is available, same flow is defaulted)
works as usual. Takes the onnx_importer route -> LoadNetwork() -> Inference_Engine

**OPTION 2: (optional)** -> Enabled via **Runtime options** via API's
so, in this flow, you can switch on and off import blob mode using the c++ and python API's.
A new runtime option has been added with name **use_compiled_network**, which could be set to true, to enable this
**Enabling save/Load blob** feature.

**2a.**
once a model is passed to OpenVINO-EP, it tries to search for it's corresponding device_specific compiled blob. If found, it will go directly via import Network flow where it reads the blob from the compiled_blobs folder. (Excludes the usual onnx importer route).

**2b.**
If blob not found, it will take the usual onnx_importer route and it will dump a compiled blob from exe_network once the model is loaded to the device specific plugin.
so, next time, the same model is passed, it will pick it's corresponding blob and works fine.

**so, the naming convention, I am following to save a blob, so that the blob is unique all the time is:**

model_name (from graph_viewer object) + opset version (from graph_viewer object) + name of the actual model passed by user which I scrap from the model path (from graph_viewer object) + hw_target_name + full / partial + .blob

Note: I had to use this set of combinations for naming to make sure the blob name is unique every time.

**OPTION 3: (optional)**
say, a user wants to directly pass in a already compiled blob directly to the OpenVINO-EP.
The user has to follow these steps. (Has to switch on two environment variables)
1.
export OV_BLOB_PATH='/home/maajid/project_files/keembay_maajid_mar1/onnxruntime/build/Linux/Debug/compiled_blobs/'

2.
**export OV_USE_COMPILED_NETWORK=1**

This way the EP will directly pick the compiled blob and take the import Network route.

### CODE REVIEW

**Comment by MaajidKhan** [comment]
**I need some more info w.r.t to this PR:**

1.I couldn't update the OpenVINO-EP documentation page yet. Looks like the file where we used to update the docx has been removed. can I know, how can I update the docx for the EP.  Any Location?

2.The **OPTION 3** from the PR description is optional. Let me know if this is also required or you want me to remove this extra option. This option is for users who directly want to infer directly using the compiled blob (blob available at any place in the disk).

3. 
currently in **OPTION 2**, where we enable save/load feature. The openvino-EP creates a directory named **ov_compiled_blobs** using cmake during build time. This directory would be dumped at a fixed location. For example with Release build, it would be dumped at **/onnxruntime/build/Linux/Release** .

All the blobs would be saved here and also picked from here. I felt this location would be ideal. same applies for windows as well.

**Comment by HectorSVC** [comment]
> **I need some more info w.r.t to this PR:**
> 
> 1.I couldn't update the OpenVINO-EP documentation page yet. Looks like the file where we used to update the docx has been removed. can I know, how can I update the docx for the EP. Any Location?
> 
> 2.The **OPTION 3** from the PR description is optional. Let me know if this is also required or you want me to remove this extra option. This option is for users who directly want to infer directly using the compiled blob (blob available at any place in the disk).
> 
> 
> currently in **OPTION 2**, where we enable save/load feature. The openvino-EP creates a directory named **ov_compiled_blobs** using cmake during build time. This directory would be dumped at a fixed location. For example with Release build, it would be dumped at **/onnxruntime/build/Linux/Release** .
> 
> All the blobs would be saved here and also picked from here. I felt this location would be ideal. same applies for windows as well.

The docs are moved to here: https://www.onnxruntime.ai/docs/get-started/.  Click the "edit this page on Github" at the bottom, it will lead you to another Github: https://github.com/microsoft/onnxruntime/blob/gh-pages/docs/get-started/index.md.

**Comment by MaajidKhan** [comment]
Added two new functionalities to this feature as requested by George @jywu-msft 
 
1.The 'ov_compiled_blobs' directory won't be dumped by cmake anymore during build time instead it would be dumped at the run time at the executable path automatically. Also additionally, different checks have been added properly added to throw out an error wherever required and provide necessary info to the user what is happening when they run an application at run time.

2.Also, Extra Run time option has been introduced via C/python API's to explicitly configure the path for save/load blob. when this option is exercised, it over rides the default executable path.

**Example:**

./onnxruntime_perf_test -r 1 -c 1 -v -I **-e openvino** **-i "use_compiled_network|true blob_dump_path|"/home/intel/maajid/project_files/save_load_blob_mar23/intel_onnxrt_myriad_fp16_ov_2021.2_binary/onnxruntime/build/Linux/Release/maajid""** /home/intel/maajid/project_files/data_models_onnx_and_tests/alexnet/model.onnx

Feature changes verified for both Linux and windows.

**Comment by MaajidKhan** [comment]
There's an issue just with the **python API** for explicitly changing the save/load dump location using run_time option.
Debug is in progress.

**Comment by jywu-msft** [comment]
> There's an issue just with the **python API** for explicitly changing the save/load dump location using run_time option.
> Debug is in progress.

Thanks for the update @MaajidKhan 
From our side, bear with us, we will need more time to review this and also try it out for ourselves.

File: onnxruntime/core/providers/openvino/backend_utils.cc [hunk 1]
```diff
@@ -40,6 +40,60 @@ void DumpOnnxModelProto(const ONNX_NAMESPACE::ModelProto& model_proto, std::stri
 
 #endif
 
+bool UseCompiledNetwork() {
+#ifdef _WIN32
+  size_t env_name_len = 0;
+  char* env_name = nullptr;
+  bool res = (_dupenv_s(&env_name, &env_name_len, "OV_USE_COMPILED_NETWORK") == 0 && env_name != nullptr);
```
Comment (jywu-msft) [inline onnxruntime/core/providers/openvino/backend_utils.cc:47]: use onnxruntime::GetEnvironmentVar() ?

File: onnxruntime/core/providers/openvino/openvino_execution_provider.cc [hunk 2]
```diff
@@ -55,6 +58,14 @@ OpenVINOExecutionProvider::GetCapability(const GraphViewer& graph_viewer, const
   ORT_UNUSED_PARAMETER(kernel_registries);
 
   std::vector<std::unique_ptr<ComputeCapability>> result;
+  openvino_ep::BackendManager::GetGlobalContext().onnx_model_name = graph_viewer.Name();
+  #ifdef _WIN32
+	  std::wstring onnx_path = graph_viewer.ModelPath().ToPathString();
```
Comment (jywu-msft) [inline onnxruntime/core/providers/openvino/openvino_execution_provider.cc:63]: indents looks off here.

File: onnxruntime/core/providers/openvino/backends/basic_backend.cc [hunk 3]
```diff
@@ -26,56 +27,142 @@ BasicBackend::BasicBackend(const ONNX_NAMESPACE::ModelProto& model_proto,
...
-#endif
-  if (global_context_.device_type.find("MYRIAD") != std::string::npos) {
-#if defined(OPENVINO_2021_1) || defined(OPENVINO_2021_2)
-    if (subgraph_context_.set_vpu_config) {
-      config["MYRIAD_DETECT_NETWORK_BATCH"] = CONFIG_VALUE(NO);
+
+  char* compiled_blob_path = nullptr;
+  if (vpu_status == true || openvino_ep::backend_utils::UseCompiledNetwork()) {
+  #ifdef _WIN32
+		size_t env_name_len = 0;
```
Comment (jywu-msft) [inline onnxruntime/core/providers/openvino/backends/basic_backend.cc:88]: indents look off here.
make sure code conforms to .clang-format in root of repo.

**Comment by oliviajain** [comment]
Hi @MaajidKhan, I wanted to provide you with an update of the validation and I had a question. 

**Validating**
1. Option 2 - C++: 
- Saving blob to a particular location
- Loading blob from a particular place
- Saving blob to a default location
- Loading blob from a default location 
2. Option 2 - Python **Need to do**
3. Option 3 - Environment Variables
- I tried loading from environment variables but got an exception. I see it was optional above, but it looks supported in the code. Let me know if you want the full log, or I tested it incorrectly. 
- 
**export OV_USE_COMPILED_NETWORK=1**
**export  OV_BLOB_PATH='/home/upsquared/olivia/onnxruntime/build/Linux/Release/ov_compiled_blobs/'**
**./onnxruntime_perf_test -r 3 -m times -c 1 -v -I -e openvino /home/upsquared/olivia/onnxruntime/onnxruntime/test/testdata/mnist.onnx r.txt**
2021-04-06 17:27:51.833609253 [I:onnxruntime:Default, basic_backend.cc:93 BasicBackend] [OpenVINO-EP] Importing the pre-compiled blob from the path set by the user
2021-04-06 17:27:53.685182522 [E:onnxruntime:, inference_session.cc:1305 operator()] Exception during initialization: /home/upsquared/olivia/onnxruntime/onnxruntime/core/providers/openvino/backends/basic_backend.cc:99 onnxruntime::openvino_ep::BasicBackend::BasicBackend(const onnx::ModelProto&, onnxruntime::openvino_ep::GlobalContext&, const onnxruntime::openvino_ep::SubGraphContext&) [OpenVINO-EP]  **Exception while Importing Network for graph: OpenVINOExecutionProvider_OpenVINO-EP-subgraph_1_0: std::bad_alloc**

**Comment by MaajidKhan** [comment]
> ./onnxruntime_perf_test -r 3 -m times -c 1 -v -I -e openvino

@oliviajain 
For Option 3, you need to set the whole path of the model instead of model directory path. since this option picks the already compiled blob and does the inference.

something like this in your case.
export OV_USE_COMPILED_NETWORK=1
export OV_BLOB_PATH='/home/upsquared/olivia/onnxruntime/build/Linux/Release/ov_compiled_blobs/**mnist.blob**'

Remember to test out option 3 using onnxruntime_perf_test tool, you need to run the same model for which the blob path is already set using the environment variables.

**For example:**
say you have a compiled blob (.blob) file for mnist model (mnist.blob) at /home/upsquared/olivia/onnxruntime/build/Linux/Release/ov_compiled_blobs/

Then test out option 3, something like this.
export OV_USE_COMPILED_NETWORK=1
export OV_BLOB_PATH='/home/upsquared/olivia/onnxruntime/build/Linux/Release/ov_compiled_blobs/**mnist.blob**'

./onnxruntime_perf_test -r 3 -m times -c 1 -v -I -e openvino /home/upsquared/olivia/onnxruntime/onnxruntime/test/testdata/mnist.onnx

**Comment by oliviajain** [comment]
Two Questions: 
1. I'm following your instructions about setting the path to the blob in the environment variable, and am able to run without an exception but I don't see any [logs ](https://github.com/microsoft/onnxruntime/blob/cd0ec5ea107b7effc95a8bfc9c633b44b779f20b/onnxruntime/core/providers/openvino/backends/basic_backend.cc#L85) about blobs which I expect to see. 
2. I am unable to build the python wheel properly: 
At first I was getting cmake errors so I added cmake_extra_defines to the build command ** ./build.sh --config Release --use_openvino MYRIAD_FP16 --parallel --skip_tests --build_shared_lib --build_wheel --cmake_extra_defines PYTHON_INCLUDE_DIR=/usr/include/python3.6m PYTHON_LIBRARY=/usr/lib/x86_64-linux-gnu/librt.so
** what's your build command? When running using session options I get this: 
upsquared@Up2-Edge-02:~/olivia$ python3 test.py
Traceback (most recent call last):
  File "test.py", line 26, in <module>
    main()
  File "test.py", line 9, in main
    options.use_compiled_network=True
**AttributeError: 'onnxruntime.capi.onnxruntime_pybind11_state.SessionOptions' object has no attribute 'use_compiled_network'**

Maybe let's sync up?

**Comment by oliviajain** [comment]
Validating

Option 2 - C++: **Done**
Saving blob to a particular location
Loading blob from a particular place
Saving blob to a default location
Loading blob from a default location
Option 2 - Python **Done**
Saving blob to a particular location
Loading blob from a particular place
Saving blob to a default location
Loading blob from a default location
Option 3 - Environment Variables **Done**
Saving blob to a particular location
Loading blob from a particular place
Getting Exception if BLOB_PATH empty - yes

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**